src/shogun/mathematics/linalg/LinalgBackendViennaclKernels.h

```diff
@@ -41,6 +41,204 @@
 
 namespace shogun
 {
+	/** Generates the computation kernel for convolution with a stride of 1
+	* The OpenCL kernel that helps to calculate the convolution or cross-correlation
+	* of SGMatrices with strides of 1
+	*/
+	template <class T>
+	static viennacl::ocl::kernel& generate_kernel_unity_stride(
+		int32_t radius_x, int32_t radius_y, bool flip)
+	{
+		std::string kernel_name =
+			"convolve_unity_stride_" + linalg::implementation::ocl::get_type_string<T>()
+			+ "_" + std::to_string(radius_x) + "_" + std::to_string(radius_y);
+
+		if (flip) kernel_name.append("_flip");
+
+		if (linalg::implementation::ocl::kernel_exists(kernel_name))
+			return linalg::implementation::ocl::get_kernel(kernel_name);
+
+		std::string source = linalg::implementation::ocl::
+				generate_kernel_preamble<T>(kernel_name);
+
+		if (flip) source.append("#define FLIP\n");
+
+		source.append("#define RADIUS_X " + std::to_string(radius_x) + "\n");
+		source.append("#define RADIUS_Y " + std::to_string(radius_y) + "\n");
+
+		source.append(
+			R"(
+				#define W_WIDTH (2*RADIUS_X+1)
+				#define W_HEIGHT (2*RADIUS_Y+1)
+
+				#define X_LOCAL_WIDTH (WORK_GROUP_SIZE_2D+2*RADIUS_X)
+				#define X_LOCAL_HEIGHT (WORK_GROUP_SIZE_2D+2*RADIUS_Y)
+
+				inline DATATYPE readX(read_only __global DATATYPE* X, int x, int y,
+					int X_width, int X_height, int X_offset)
+				{
+					if (x>=0 && y>=0 && x<X_width && y<X_height)
+						return X[y + x*X_height + X_offset];
+					else
+						return 0;
+				}
+
+				__kernel void KERNEL_NAME(
+					read_only __global DATATYPE* X, int X_width, int X_height, int X_offset,
+					__constant DATATYPE* W, int W_offset,
+					__global DATATYPE* Y, int Y_offset)
+				{
+					__local DATATYPE X_local[X_LOCAL_WIDTH][X_LOCAL_HEIGHT];
+
+					int x = get_global_id(0);
+					int y = get_global_id(1);
+
+					int xl = get_local_id(0);
+					int yl = get_local_id(1);
+
+					if (xl==WORK_GROUP_SIZE_2D-1 && yl == WORK_GROUP_SIZE_2D-1)
+					{
+						for (int rx=0; rx<=2*RADIUS_X; rx++)
+							for (int ry=0; ry<=2*RADIUS_Y; ry++)
+								X_local[xl+rx][yl+ry] = readX(X, x-RADIUS_X+rx, y-RADIUS_Y+ry, X_width, X_height, X_offset);
+					}
+					else if (xl==WORK_GROUP_SIZE_2D-1)
+					{
+						for (int rx=0; rx<=2*RADIUS_X; rx++)
+							X_local[xl+rx][yl] = readX(X, x-RADIUS_X+rx, y-RADIUS_Y, X_width, X_height, X_offset);
+					}
+					else if (yl == WORK_GROUP_SIZE_2D-1)
```

Comment:
This is crazy code
Did the original state where it's coming from? Is it tested? Used?